cvat/apps/engine/models.py

```diff
@@ -181,6 +183,242 @@ def choices(cls):
+                self._mark_parsed()
+
+            element_str = string[current_position:end]
+            current_position = end + separator_offset
+            if not element_str:
+                self._probable_length -= 1
+                continue
+            element = self._converter(element_str)
+            if list.__len__(self) <= current_index:
+                # We need to handle special case when instance of lazy list becomes parsed after
+                # this function is called:
+                # ll = LazyList("1,2,3", _converter=int)
+                # iterator = iter(ll)
+                # next(iterator)  # > 1 (will generate next element and append to self)
+                # list(ll)  # > [1, 2, 3]
+                # next(iterator)  # > 2 (will generate next element, however will not append it)
+                # assert list(ll) == [1, 2, 3]
+                list.append(self, element)
+            yield element
+            current_index += 1
+
+    def _compute_max_length(self, string) -> int:
+        if self._probable_length is None:
+            if not self.string:
+                return 0
+            self._probable_length = string.count(self._separator) + 1
+        return self._probable_length
+
+    # support pickling
+
+    def __reduce__(self):
+        return self.__class__, (self.string, self._separator, self._converter), self.__getstate__()
+
+    def __reduce_ex__(self, protocol: int):
+        return self.__reduce__()
+
+    def __getstate__(self):
+        return {
+            'string': self.string,
+            '_separator': self._separator,
+            '_converter': self._converter,
+            '_probable_length': self._probable_length,
+            'parsed': self.parsed,
+            'parsed_elements': list(self) if self.parsed else None
+        }
+
+    def __setstate__(self, state):
+        self.string = state['string']
+        self._separator = state['_separator']
+        self._converter = state['_converter']
+        self._probable_length = state['_probable_length']
+        self.parsed = state['parsed']
+        if self.parsed:
+            self.extend(state['parsed_elements'])
+
+    # add pre-parse for list methods
+    for method in [
+        "append",
+        "copy",
+        "insert",
+        "pop",
+        "remove",
+        "reverse",
+        "sort",
+        "clear",
+        "index",
+        "count",
+        "__setitem__",
+        "__delitem__",
+        "__contains__",
+        "__len__",
+        "__contains__",
+        "__reversed__",
+        "__mul__",
+        "__rmul__",
+        "__imul__",
+    ]:
+        locals()[method] = _parse_before_accessing(getattr(list, method))
+
+    for method in [
+        "extend",
+        "__add__",
+        "__eq__",
+        "__iadd__",
+        "__gt__",
+        "__ge__",
+        "__lt__",
+        "__le__",
+        "__eq__",
+    ]:
+        locals()[method] = _parse_both_before_accessing(getattr(list, method))
```

Comment:
Modifying locals [is UB](https://docs.python.org/3/library/functions.html#locals) in python. Typically, it can be replaced either by overriding specific methods, `__getattribute__`, or a metaclass. One more way is to create the class in a function, which can also write some fields later.

Reply:
Though in this particular case I didn't see any complications, you're right, `locals()` might be too fragile to rely on.

Using `__getattribute__`, on the other hand, would introduce all sorts of complications, so I went with moving this code to metaclass.